jax/core.py

```diff
@@ -1219,19 +1226,24 @@ def write(v: Var, a: AbstractValue) -> None:
   map(write, jaxpr.constvars, [v.aval for v in jaxpr.constvars])
   map(write, jaxpr.invars, in_avals)
 
-  for eqn in jaxpr.eqns:
+  for eqn_idx, eqn in enumerate(jaxpr.eqns):
     in_avals = map(read, eqn.invars)
-    if eqn.primitive.call_primitive:
-      out_avals = check_call(eqn.primitive, in_avals, eqn.params)
-    elif eqn.primitive.map_primitive:
-      out_avals = check_map(eqn.primitive, in_avals, eqn.params)
-    else:
-      out_avals = check_eqn(eqn.primitive, in_avals, eqn.params)
+    prim = eqn.primitive
     try:
+      if prim in custom_typechecks:
+        custom_typechecks[prim](*in_avals, **eqn.params)
```

Comment:
Should this be like `out_avals = custom_typechecks[prim](...)` and an `elif` on the next line? Then the custom typecheck rule can avoid the standard typecheck rule (though it can always turn around and call its abstract eval if desired).

Reply:
That would allow the typechecking and abstract evaluation rules to differ. The abstract eval rule determines behavior outside of the typechecker, e.g. in lowering. For now, I'd rather it be clear that what we deduce by typechecking matches reality elsewhere. Because I don't have an example in mind where "custom abstract eval only when typechecking" would be useful, I'd rather wait to change this as needed. There are more typechecking rules to write after the two in this PR, so we'll have a chance to find out and revise.